internal/common.go

```diff
@@ -56,3 +62,65 @@ const (
 	// ShimBinaryName is the name of the runtime shim binary
 	ShimBinaryName = "containerd-shim-aws-firecracker"
 )
+
+// MagicStubBytes used to determine whether or not a drive is a stub drive
+var MagicStubBytes = []byte{214, 244, 216, 245, 215, 177, 177, 177}
+
+// IsStubDrive will check to see if the io.Reader follows our stub drive
+// format. In the event of an error, this will return false. This will only
+// return true when the first n bytes read matches that of the magic stub
+// bytes and where n is the length of magic bytes.
+func IsStubDrive(r io.Reader) bool {
+	buf := make([]byte, len(MagicStubBytes))
+	lr := io.LimitReader(r, int64(len(MagicStubBytes)))
+	n, err := lr.Read(buf)
+	if err != nil {
+		return false
+	}
+
+	if n != len(MagicStubBytes) {
+		return false
+	}
+
+	return bytes.Equal(buf[:n], MagicStubBytes)
+}
+
+// ParseStubContent .
+func ParseStubContent(r io.Reader) (string, error) {
+	magicBytesReader := io.LimitReader(r, int64(len(MagicStubBytes)))
+	magicBytes := make([]byte, len(MagicStubBytes))
+	_, err := magicBytesReader.Read(magicBytes)
+	if err != nil {
+		return "", err
+	}
+
+	sizeReader := io.LimitReader(r, 1)
+	sizeByte := make([]byte, 1)
+	_, err = sizeReader.Read(sizeByte)
+	if err != nil {
+		return "", err
+	}
+
+	idReader := io.LimitReader(r, int64(sizeByte[0]))
+	idBytes := make([]byte, sizeByte[0])
+	_, err = idReader.Read(idBytes)
+	if err != nil {
+		return "", err
+	}
+
+	return string(idBytes), nil
+}
+
+// GenerateStubContent .
+func GenerateStubContent(w io.Writer, id string) error {
+	length := len(id)
+	if length > 0xFF {
```

Comment:
The length of the drive ID is limited to 255 bytes, except that it needs to be smaller to accommodate the `MagicStubBytes`.

Should be:

```go
const maximumStubContentLength = 255
if length + len(MagicStubBytes) > maximumStubContentLength
```

Reply:
@samuelkarp - I put the magic bytes first. So the expected format is like so `<magic bytes><size><drive id>`. This excludes the size of the magic stub bytes and do not need to calculated.

Reply:
You're right; I was confusing this with the maximum stub drive content.  We need to verify somewhere that the stub content doesn't exceed 512 blocks, since any content beyond that gets truncated.  That can either happen inside StubDrivePaths or GenerateStubContent.